solidity/contracts/Coordinator.sol

```diff
@@ -326,12 +305,28 @@ contract Coordinator is ChainlinkRequestInterface, CoordinatorInterface {
     isValidRequest(_requestId)
     returns (bool)
   {
-    Callback memory callback = callbacks[_requestId];
-    delete callbacks[_requestId];
-    // All updates to the oracle's fulfillment should come before calling the
-    // callback(addr+functionId) as it is untrusted. See:
-    // https://solidity.readthedocs.io/en/develop/security-considerations.html#use-the-checks-effects-interactions-pattern
-    return callback.addr.call(callback.functionId, _requestId, _data); // solium-disable-line security/no-low-level-calls
+    bytes32 requestId = bytes32(_requestId);
+    callbacks[requestId].responses[msg.sender] = uint256(_data);
+
+    Callback memory callback = callbacks[requestId];
+    callbacks[requestId].responseCount += 1;
+    uint256 responseCount = callbacks[requestId].responseCount;
+    bytes32 sAId = callbacks[requestId].sAId;
+    address[] memory oracles = serviceAgreements[sAId].oracles;
+    if (oracles.length > responseCount) {
+      return true;
+    }
+
+    uint256 result;
+    uint256 oraclePayment = callback.amount.div(oracles.length);
+    for (uint i = 0; i < responseCount; i++) {
+      result = result.add(callbacks[requestId].responses[oracles[i]]);
+      delete callbacks[requestId].responses[oracles[i]];
+      withdrawableTokens[oracles[i]] = withdrawableTokens[oracles[i]].add(oraclePayment);
+    }
+    result = result.div(responseCount);
+    delete callbacks[requestId];
```

Comment:
SafeMath's `.div()` operation has no assertion:
```
  /**
  * @dev Integer division of two numbers, truncating the quotient.
  */
  function div(uint256 _a, uint256 _b) internal pure returns (uint256) {
    // assert(_b > 0); // Solidity automatically throws when dividing by 0
    // uint256 c = _a / _b;
    // assert(_a == _b * c + _a % _b); // There is no case in which this doesn't hold
    return _a / _b;
  }
```

Reply:
SaveMath won't, but it looks like solidity will?

Reply:
In what case would aggregation run with 0 responses, which is where we would be dividing by 0?

Reply:
Not sure, just pointing out that even though SaveMath has no assertion, div can still fail.

Reply:
So basically how I see it is the fix for this is to use SafeMath for the addition operation on line 235, incrementing the responseCount. Just checking, are we good with that?

Reply:
Following up myself on this and I think we should make this change. The responseCount variable now is a uint8 (I had assumed it was already uint256), and we're not checking that some max number of oracles is being sent in for the initiateServiceAgreement function. So, assuming we didn't hit the gas limit for a block, we could overflow with 256 oracles, which SafeMath would guard against. This would also require changing responseCount in the Callback struct to uint256.